File: web/src/shared/deck.ts
import { Card, Rank, Suit } from './types';

export function createDeck(): Card[] {
  const cards: Card[] = [];
  const suits = [Suit.Spades, Suit.Hearts, Suit.Clubs, Suit.Diamonds];
  const ranks = [
    Rank.Two, Rank.Three, Rank.Four, Rank.Five, Rank.Six, Rank.Seven,
    Rank.Eight, Rank.Nine, Rank.Ten, Rank.Jack, Rank.Queen, Rank.King, Rank.Ace
  ];

  // Two decks
  for (let i = 0; i < 2; i++) {
    // Standard cards
    for (const suit of suits) {
      for (const rank of ranks) {
        cards.push({
          suit,
          rank,
          id: `${suit}-${rank}-${i}`
        });
      }
    }
    // Jokers
    cards.push({ suit: Suit.Joker, rank: Rank.SmallJoker, id: `joker-small-${i}` });
    cards.push({ suit: Suit.Joker, rank: Rank.BigJoker, id: `joker-big-${i}` });
  }

  return cards;
}

export function shuffleDeck(cards: Card[]): Card[] {
  for (let i = cards.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [cards[i], cards[j]] = [cards[j], cards[i]];
  }
  return cards;
}

// Helper to update card properties based on current game level
export function updateCardProperties(cards: Card[], currentLevel: number): Card[] {
  return cards.map(card => {
    const isLevelCard = card.rank === currentLevel;
    const isWild = isLevelCard && card.suit === Suit.Hearts;
    return { ...card, isLevelCard, isWild };
  });
}

File: internal/game/deck.go
package game

import (
	"fmt"
	"math/rand"
)

// CreateDeck builds two full decks (108 cards) — mirrors shared/deck.ts createDeck.
func CreateDeck() []Card {
	cards := make([]Card, 0, 108)
	suits := []Suit{Spades, Hearts, Clubs, Diamonds}

	for i := 0; i < 2; i++ {
		for _, suit := range suits {
			for rank := RankTwo; rank <= RankAce; rank++ {
				cards = append(cards, Card{
					Suit: suit,
					Rank: rank,
					ID:   fmt.Sprintf("%d-%d-%d", suit, rank, i),
				})
			}
		}
		cards = append(cards, Card{Suit: JokerSuit, Rank: SmallJoker, ID: fmt.Sprintf("joker-small-%d", i)})
		cards = append(cards, Card{Suit: JokerSuit, Rank: BigJoker, ID: fmt.Sprintf("joker-big-%d", i)})
	}
	return cards
}

func ShuffleDeck(cards []Card) []Card {
	rand.Shuffle(len(cards), func(i, j int) {
		cards[i], cards[j] = cards[j], cards[i]
	})
	return cards
}

// UpdateCardProperties sets isLevelCard/isWild flags based on the current level.
func UpdateCardProperties(cards []Card, currentLevel int) []Card {
	out := make([]Card, len(cards))
	for i, card := range cards {
		card.IsLevelCard = int(card.Rank) == currentLevel
		card.IsWild = card.IsLevelCard && card.Suit == Hearts
		out[i] = card
	}
	return out
}
